File: shen-derive/specfile/typetable.go
package specfile

import (
	"strings"
	"unicode"
)

// TypeCategory classifies how a Shen datatype maps to Go.
// Mirrors shengen/main.go classification at lines 161-177.
type TypeCategory string

const (
	CatWrapper     TypeCategory = "wrapper"     // single-field, no verification: struct{v T} + NewX(x) X
	CatConstrained TypeCategory = "constrained" // single-field, verified premises: NewX(x) (X, error)
	CatComposite   TypeCategory = "composite"   // multi-field, no verification
	CatGuarded     TypeCategory = "guarded"     // multi-field, verified
	CatAlias       TypeCategory = "alias"       // type X = Y
	CatSumType     TypeCategory = "sumtype"     // union produced by multiple datatypes
)

// TypeEntry describes the Go shape of one Shen type.
type TypeEntry struct {
	ShenName    string
	GoName      string // "Amount" (unqualified)
	GoQualified string // "shenguard.Amount" (with import alias)
	Category    TypeCategory

	// For wrapper/constrained:
	ShenPrim   string // "number", "string", "boolean", "symbol"
	GoPrimType string // "float64", "string", "bool"
	VarName    string // the premise variable name (e.g. "X"), used when evaluating Verified

	// For constrained and guarded: raw verified s-expressions, e.g. "(>= X 0)".
	Verified []string

	// For composite/guarded:
	Fields []Field
}

// Field is one field of a composite Shen datatype.
type Field struct {
	Index    int
	ShenName string // "Amount" (as written in the spec)
	ShenType string // "amount"
	GoMethod string // "Amount" (the Go accessor method name)
}

// TypeTable maps Shen type names to their Go equivalents.
type TypeTable struct {
	Entries     map[string]*TypeEntry
	ImportPath  string // e.g., "example.com/project/internal/shenguard"
	ImportAlias string // e.g., "shenguard"
}

// BuildTypeTable classifies every datatype and builds a type table.
// importPath is the Go import path of the shengen-generated package;
// importAlias is the short name used to qualify types (e.g., "shenguard").
func BuildTypeTable(datatypes []Datatype, importPath, importAlias string) *TypeTable {
	tt := &TypeTable{
		Entries:     make(map[string]*TypeEntry),
		ImportPath:  importPath,
		ImportAlias: importAlias,
	}

	// Count how many datatype blocks produce each conclusion type.
	// Multiple → sum type variants.
	concCount := make(map[string]int)
	for _, dt := range datatypes {
		for _, r := range dt.Rules {
			concCount[r.Conclusion.TypeName]++
		}
	}

	for _, dt := range datatypes {
		for _, r := range dt.Rules {
			typeName := r.Conclusion.TypeName
			// Sum variants use the block name; the union type itself is registered below.
			isSumVariant := dt.Name != r.Conclusion.TypeName && concCount[r.Conclusion.TypeName] > 1
			if isSumVariant {
				typeName = dt.Name
			}

			entry := &TypeEntry{
				ShenName: typeName,
				GoName:   toPascalCase(typeName),
			}
			entry.GoQualified = qualify(importAlias, entry.GoName)

			switch {
			case r.Conclusion.IsWrapped && len(r.Verified) == 0 && len(r.Premises) == 1 && isPrimitive(r.Premises[0].TypeName):
				entry.Category = CatWrapper
				entry.ShenPrim = r.Premises[0].TypeName
				entry.GoPrimType = shenPrimToGo(entry.ShenPrim)
				entry.VarName = r.Premises[0].VarName

			case r.Conclusion.IsWrapped && len(r.Verified) > 0 && len(r.Premises) >= 1 && isPrimitive(r.Premises[0].TypeName):
				entry.Category = CatConstrained
				entry.ShenPrim = r.Premises[0].TypeName
				entry.GoPrimType = shenPrimToGo(entry.ShenPrim)
				entry.VarName = r.Premises[0].VarName

			case r.Conclusion.IsWrapped && len(r.Premises) == 1 && !isPrimitive(r.Premises[0].TypeName) && !isSumVariant:
				entry.Category = CatAlias

			case !r.Conclusion.IsWrapped && len(r.Verified) > 0:
				entry.Category = CatGuarded

			default:
				entry.Category = CatComposite
			}

			// Record verified predicates on entries that have constraints.
			if entry.Category == CatConstrained || entry.Category == CatGuarded {
				for _, v := range r.Verified {
					entry.Verified = append(entry.Verified, v.Raw)
				}
			}

			// Fields for composites (and sum variants with wrapped conclusions,
			// for symmetry with shengen).
			if !r.Conclusion.IsWrapped || isSumVariant {
				premMap := map[string]string{}
				for _, p := range r.Premises {
					premMap[p.VarName] = p.TypeName
				}
				for i, fieldName := range r.Conclusion.Fields {
					shenType := premMap[fieldName]
					if shenType == "" {
						shenType = "unknown"
					}
					entry.Fields = append(entry.Fields, Field{
						Index:    i,
						ShenName: fieldName,
						ShenType: shenType,
						GoMethod: toPascalCase(fieldName),
					})
				}
			}

			tt.Entries[typeName] = entry
		}
	}

	// Register synthetic sum type entries for union conclusions.
	for _, dt := range datatypes {
		for _, r := range dt.Rules {
			conc := r.Conclusion.TypeName
			if concCount[conc] > 1 {
				if _, ok := tt.Entries[conc]; !ok {
					tt.Entries[conc] = &TypeEntry{
						ShenName:    conc,
						GoName:      toPascalCase(conc),
						GoQualified: qualify(importAlias, toPascalCase(conc)),
						Category:    CatSumType,
					}
				}
			}
		}
	}

	return tt
}

// GoType returns the Go type expression for a Shen type.
// Handles "(list T)" by recursing on the element type.
// Primitive types map to Go built-ins; declared types are qualified with the import alias.
func (tt *TypeTable) GoType(shenType string) string {
	shenType = strings.TrimSpace(shenType)
	if strings.HasPrefix(shenType, "(list ") && strings.HasSuffix(shenType, ")") {
		inner := strings.TrimSpace(shenType[len("(list ") : len(shenType)-1])
		return "[]" + tt.GoType(inner)
	}
	if isPrimitive(shenType) {
		return shenPrimToGo(shenType)
	}
	if e, ok := tt.Entries[shenType]; ok {
		return e.GoQualified
	}
	// Unknown — fall back to Pascal case, no qualification.
	return toPascalCase(shenType)
}

// ElemType extracts the element type from "(list T)" or returns "".
func ElemType(shenType string) string {
	shenType = strings.TrimSpace(shenType)
	if strings.HasPrefix(shenType, "(list ") && strings.HasSuffix(shenType, ")") {
		return strings.TrimSpace(shenType[len("(list ") : len(shenType)-1])
	}
	return ""
}

// TypeSummary represents one Shen datatype for context output.
type TypeSummary struct {
	ShenName     string   // e.g. "balance-checked"
	TargetName   string   // e.g. "BalanceChecked" (Go name)
	Category     string   // "wrapper", "constrained", "composite", "guarded", "alias", "sumtype"
	Fields       []string // field names for composite/guarded types
	Constraints  []string // raw constraint expressions for constrained/guarded types
	Dependencies []string // other Shen type names referenced
}

// Summary returns a structured overview of all types in the table.
func (tt *TypeTable) Summary() []TypeSummary {
	var out []TypeSummary
	for _, e := range tt.Entries {
		s := TypeSummary{
			ShenName:   e.ShenName,
			TargetName: e.GoName,
			Category:   string(e.Category),
		}
		for _, f := range e.Fields {
			s.Fields = append(s.Fields, f.ShenName)
		}
		s.Constraints = append(s.Constraints, e.Verified...)

		// Collect dependencies: field types that are not primitives.
		seen := map[string]bool{}
		for _, f := range e.Fields {
			if !isPrimitive(f.ShenType) && f.ShenType != "unknown" && !seen[f.ShenType] {
				s.Dependencies = append(s.Dependencies, f.ShenType)
				seen[f.ShenType] = true
			}
		}

		out = append(out, s)
	}
	// Sort for stable output.
	sortSummaries(out)
	return out
}

func sortSummaries(ss []TypeSummary) {
	for i := 1; i < len(ss); i++ {
		for j := i; j > 0 && ss[j].ShenName < ss[j-1].ShenName; j-- {
			ss[j], ss[j-1] = ss[j-1], ss[j]
		}
	}
}

// --- Helpers ---

func isPrimitive(t string) bool {
	return t == "string" || t == "number" || t == "boolean" || t == "symbol"
}

func shenPrimToGo(t string) string {
	switch t {
	case "number":
		return "float64"
	case "string", "symbol":
		return "string"
	case "boolean":
		return "bool"
	default:
		return "any"
	}
}

func toPascalCase(s string) string {
	parts := strings.FieldsFunc(s, func(r rune) bool { return r == '-' || r == '_' })
	var b strings.Builder
	for _, p := range parts {
		if len(p) == 0 {
			continue
		}
		runes := []rune(p)
		runes[0] = unicode.ToUpper(runes[0])
		b.WriteString(string(runes))
	}
	return b.String()
}

func qualify(alias, name string) string {
	if alias == "" {
		return name
	}
	return alias + "." + name
}

File: cmd/shen-derive-ts/specfile/typetable.ts
// Shen→TS type classifier. Mirrors shen-derive/specfile/typetable.go.
//
// Given the parsed datatypes from a .shen spec, classify each Shen type by
// its shape (wrapper, constrained, composite, guarded, alias, sumtype) and
// produce a lookup table used by the verifier and code generator.
//
// This file deliberately does NOT share code with cmd/shengen-ts. The two
// implementations are parallel, matching how the Go `specfile` package is
// parallel to `cmd/shengen/main.go`.

import type { Datatype } from "./parse.ts";

export type TypeCategory =
  | "wrapper"
  | "constrained"
  | "composite"
  | "guarded"
  | "alias"
  | "sumtype";

export type Field = {
  /** camelCase TS accessor name (e.g. "amount", "from") */
  name: string;
  /** original Shen variable name as written in the spec (e.g. "Amount") */
  shenName: string;
  /** Shen type name from the matching premise (e.g. "amount") */
  typeName: string;
  /** TS type string (e.g. "number", "Foo[]") */
  tsType: string;
};

export type TypeEntry = {
  shenName: string;
  /** PascalCase class name (e.g. "AccountId") */
  tsName: string;
  category: TypeCategory;
  /** Underlying primitive Shen type for wrappers/constrained; "" otherwise. */
  shenPrim: string;
  /** TS type string for the underlying primitive; "" otherwise. */
  tsPrimType: string;
  /** Raw verified predicate strings, e.g. "(>= X 0)". */
  verified: string[];
  /** Variable name used in verified predicates (e.g. "X"). */
  varName: string;
  /** Fields for composite/guarded types. Empty otherwise. */
  fields: Field[];
  /** Import specifier for the shengen-ts generated module. */
  importPath: string;
  /** Namespace alias used when qualifying imported types. */
  importAlias: string;
};

export type TypeTable = Map<string, TypeEntry>;

// --- Public API ---

/**
 * Classify every datatype rule and return a table keyed by Shen type name.
 * When multiple rules share a conclusion type, a synthetic `sumtype` entry is
 * registered under that conclusion name and the individual rules are stored
 * under their block names as sum variants.
 */
export function buildTypeTable(
  datatypes: Datatype[],
  importPath: string,
  importAlias: string,
): TypeTable {
  const tt: TypeTable = new Map();

  // Pass 1: count how many rules produce each conclusion type name. A count
  // greater than one means "this conclusion is a sum type with N variants".
  const concCount = new Map<string, number>();
  for (const dt of datatypes) {
    for (const r of dt.rules) {
      concCount.set(
        r.conclusion.typeName,
        (concCount.get(r.conclusion.typeName) ?? 0) + 1,
      );
    }
  }

  // Pass 2: classify each rule into an entry.
  for (const dt of datatypes) {
    for (const r of dt.rules) {
      let typeName = r.conclusion.typeName;
      const isSumVariant =
        dt.name !== r.conclusion.typeName &&
        (concCount.get(r.conclusion.typeName) ?? 0) > 1;
      if (isSumVariant) {
        typeName = dt.name;
      }

      const entry: TypeEntry = {
        shenName: typeName,
        tsName: toPascalCase(typeName),
        category: "composite",
        shenPrim: "",
        tsPrimType: "",
        verified: [],
        varName: "",
        fields: [],
        importPath,
        importAlias,
      };

      const prems = r.premises;
      const verified = r.verified;
      const conc = r.conclusion;
      // "wrapped" conclusions have the shape `X : type` (single LHS
      // variable); tuple-shaped conclusions `[A B C] : type` populate
      // `fields` instead.
      const isWrapped = conc.fields.length === 0;

      if (
        isWrapped &&
        verified.length === 0 &&
        prems.length === 1 &&
        isPrimitive(prems[0].typeName)
      ) {
        entry.category = "wrapper";
        entry.shenPrim = prems[0].typeName;
        entry.tsPrimType = shenPrimToTs(entry.shenPrim);
        entry.varName = prems[0].varName;
      } else if (
        isWrapped &&
        verified.length > 0 &&
        prems.length >= 1 &&
        isPrimitive(prems[0].typeName)
      ) {
        entry.category = "constrained";
        entry.shenPrim = prems[0].typeName;
        entry.tsPrimType = shenPrimToTs(entry.shenPrim);
        entry.varName = prems[0].varName;
      } else if (
        isWrapped &&
        prems.length === 1 &&
        !isPrimitive(prems[0].typeName) &&
        !isSumVariant
      ) {
        entry.category = "alias";
      } else if (!isWrapped && verified.length > 0) {
        entry.category = "guarded";
      } else {
        entry.category = "composite";
      }

      if (entry.category === "constrained" || entry.category === "guarded") {
        for (const v of verified) {
          entry.verified.push(v.raw);
        }
      }

      // Fields for composites (and sum variants with wrapped conclusions,
      // for symmetry with shengen).
      if (!isWrapped || isSumVariant) {
        const premMap = new Map<string, string>();
        for (const p of prems) {
          premMap.set(p.varName, p.typeName);
        }
        for (const fieldName of conc.fields) {
          const shenType = premMap.get(fieldName) ?? "unknown";
          entry.fields.push({
            name: toCamelCase(fieldName),
            shenName: fieldName,
            typeName: shenType,
            tsType: "", // filled in after the table is fully populated
          });
        }
      }

      tt.set(typeName, entry);
    }
  }

  // Pass 3: register synthetic sumtype entries for conclusions shared by
  // multiple rules.
  for (const dt of datatypes) {
    for (const r of dt.rules) {
      const conc = r.conclusion.typeName;
      if ((concCount.get(conc) ?? 0) > 1 && !tt.has(conc)) {
        tt.set(conc, {
          shenName: conc,
          tsName: toPascalCase(conc),
          category: "sumtype",
          shenPrim: "",
          tsPrimType: "",
          verified: [],
          varName: "",
          fields: [],
          importPath,
          importAlias,
        });
      }
    }
  }

  // Pass 4: now that every declared type is in the table, resolve field
  // tsType strings. This is deferred so that forward references resolve.
  for (const entry of tt.values()) {
    for (const f of entry.fields) {
      f.tsType = tsType(tt, f.typeName);
    }
  }

  return tt;
}

/**
 * Return the TS type expression for a Shen type. Handles `(list T)` by
 * recursing on the element type. Primitives map to TS built-ins; declared
 * types resolve to their PascalCase class name.
 */
export function tsType(tt: TypeTable, shenType: string): string {
  const t = shenType.trim();
  if (t.startsWith("(list ") && t.endsWith(")")) {
    const inner = t.slice("(list ".length, t.length - 1).trim();
    return tsType(tt, inner) + "[]";
  }
  if (isPrimitive(t)) {
    return shenPrimToTs(t);
  }
  const e = tt.get(t);
  if (e !== undefined) {
    return e.tsName;
  }
  return toPascalCase(t);
}

/**
 * Extract the element type from `(list T)`. Returns `""` if `shenType` is not
 * a list.
 */
export function elemType(shenType: string): string {
  const t = shenType.trim();
  if (t.startsWith("(list ") && t.endsWith(")")) {
    return t.slice("(list ".length, t.length - 1).trim();
  }
  return "";
}

// --- Helpers (ported fresh; not shared with shengen-ts) ---

function isPrimitive(t: string): boolean {
  return t === "string" || t === "number" || t === "boolean" || t === "symbol";
}

function shenPrimToTs(t: string): string {
  switch (t) {
    case "number":
      return "number";
    case "string":
    case "symbol":
      return "string";
    case "boolean":
      return "boolean";
    default:
      return "unknown";
  }
}

function splitIdent(s: string): string[] {
  return s.split(/[-_]/).filter((p) => p.length > 0);
}

function toPascalCase(s: string): string {
  return splitIdent(s)
    .map((p) => p[0].toUpperCase() + p.slice(1))
    .join("");
}

function toCamelCase(s: string): string {
  const parts = splitIdent(s);
  if (parts.length === 0) return "";
  const [head, ...rest] = parts;
  return (
    head[0].toLowerCase() +
    head.slice(1) +
    rest.map((p) => p[0].toUpperCase() + p.slice(1)).join("")
  );
}
